crates/wasmparser/src/validator/types.rs

```diff
@@ -2466,93 +2466,104 @@ impl TypeList {
-            let len = range.end.index() - range.start.index();
-            if len != rec_group.types().len() {
-                return false;
-            }
+        let start = self.core_types.len();
+        let start = u32::try_from(start).unwrap();
+        let start = CoreTypeId::from_index(start);
 
-            (range.start.index()..range.end.index())
-                .map(|i| &self.core_types[i])
-                .zip(rec_group.types())
-                .all(|(canon_ty, new_ty)| canon_ty == new_ty)
-        });
+        for ty in entry.key().types() {
+            debug_assert_eq!(self.core_types.len(), self.core_type_to_supertype.len());
+            debug_assert_eq!(self.core_types.len(), self.core_type_to_rec_group.len());
 
-        let (is_new, occupied_entry) = match entry {
-            // Occupied: use the existing entry.
-            Ok(entry) => (false, entry),
-
-            // Absent: intern the types, record their range, add a new canonical
-            // rec group for that range, insert it into the hash table, and
-            // return the new entry.
-            Err(absent_entry) => {
-                let table = absent_entry.into_table();
-
-                let rec_group_id = self.rec_group_elements.len();
-                let rec_group_id = u32::try_from(rec_group_id).unwrap();
-                let rec_group_id = RecGroupId::from_index(rec_group_id);
-
-                let start = self.core_types.len();
-                let start = u32::try_from(start).unwrap();
-                let start = CoreTypeId::from_index(start);
-
-                for ty in rec_group.into_types() {
-                    debug_assert_eq!(self.core_types.len(), self.core_type_to_supertype.len());
-                    debug_assert_eq!(self.core_types.len(), self.core_type_to_rec_group.len());
-
-                    self.core_type_to_supertype.push(ty.supertype_idx.map(
-                        |idx| match idx.unpack() {
-                            UnpackedIndex::RecGroup(offset) => {
-                                CoreTypeId::from_index(start.index + offset)
-                            }
-                            UnpackedIndex::Id(id) => id,
-                            UnpackedIndex::Module(_) => unreachable!("in canonical form"),
-                        },
-                    ));
-                    self.core_types.push(ty);
-                    self.core_type_to_rec_group.push(rec_group_id);
-                }
+            self.core_type_to_supertype
+                .push(ty.supertype_idx.map(|idx| match idx.unpack() {
+                    UnpackedIndex::RecGroup(offset) => CoreTypeId::from_index(start.index + offset),
+                    UnpackedIndex::Id(id) => id,
+                    UnpackedIndex::Module(_) => unreachable!("in canonical form"),
+                }));
+            let mut ty = ty.clone();
+            canonicalize_sub_type(&mut ty, start);
```

Comment:
This doesn't make sense to me -- it should already be in canonical form when we call this function? Otherwise the hash map lookup makes no sense in the first place.

Reply:
We talked about this but I wanted to write this down here for others too, but this is changing `RecGroup` canonicalization into `Id` canonicalization. I'll work on getting this to share code with the preexisting mapping of indices.